Issue #10956 - Send Reset NO_ERROR for HTTP/2

gregw opened the pull request:
For #10956

gregw committed:
Reviewed Expect 100 Continue Handler for #10956

Removed duplicate handling from servlet. Now only handled in core on first demand.
Added additional tests.

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12/10956/Expect100Continue

gregw committed:
Updates from review

added the willRead method to HttpStream and Request to allow signal that a read will be attempted.

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12/10956/Expect100Continue

gregw committed:
Send reset NO_ERROR for consumeAvailable in H2

gregw committed:
Send reset NO_ERROR for consumeAvailable in H2

gregw commented:
@sbordet interestingly this is passing tests.  But I think there are some exceptions (that I am printing) that we should avoid.

@lorban Can you update this branch with the same kind of changes for h3 ?

gregw committed:
Send reset NO_ERROR for consumeAvailable in H2

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12/10956/Expect100Continue

gregw committed:
Merge branch 'fix/jetty-12/10956/Expect100Continue' into fix/jetty-12/110956/Reset_NO_ERROR

lorban commented on jetty-core/jetty-http3/jetty-http3-server/src/main/java/org/eclipse/jetty/http3/server/internal/HttpStreamOverHTTP3.java:
```diff
@@ -133,6 +133,16 @@ private void onBadMessage(HttpException x)
+        if (expects100Continue)
+        {
+            expects100Continue = false;
+            send(requestMetaData, HttpGenerator.CONTINUE_100_INFO, false, null, Callback.NOOP);
```
Now this makes me wonder if this isn't racing with a user-generated write? I mean such handler:

```java
boolean handle(Request req, Response resp, Callback callback)
{
  req.demand(() -> {}); // this triggers a write with a NOOP callback
  resp.write(true, BufferUtil.EMPTY, callback); // doesn't this risk a WritePendingException?
}
```

sbordet replied:
I was thinking the same.

gregw replied:
@review that in the other PR, this PR is based on #10957, so you are seeing both changes here.

lorban commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -254,6 +254,14 @@ static long getTimeStamp(Request request)
+     * Signal the intention to read the request content for the purposes of 100-Continues processing.
+     * @see HttpStream#willRead()
+     */
+    default void willRead()
```
I think this shouldn't be marked as `default`.

gregw replied:
ditto

lorban commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/HttpStream.java:
```diff
@@ -46,6 +46,22 @@ public interface HttpStream extends Callback
+     *     state to not expect the request content to be sent and to ignore any further content sent on the channel</li>
+     * </ul>
+     */
+    default void willRead()
```
I think this shouldn't be marked as `default`.

gregw replied:
ditto

lorban commented on jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/HttpClientContinueTest.java:
```diff
@@ -111,6 +112,68 @@ protected void service(HttpServletRequest request, HttpServletResponse response)
+        testExpect100ContinueRespond100ContinueBlocking(transport, "data1".getBytes(StandardCharsets.UTF_8), "data2".getBytes(StandardCharsets.UTF_8), "data3".getBytes(StandardCharsets.UTF_8));
+    }
+
+    private void testExpect100ContinueRespond100ContinueBlocking(Transport transport, byte[]... contents) throws Exception
```
This private method looks like it could be inlined in the test method.

lorban commented on jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/HttpClientContinueTest.java:
```diff
@@ -111,6 +112,68 @@ protected void service(HttpServletRequest request, HttpServletResponse response)
+
+    private void testExpect100ContinueRespond100ContinueBlocking(Transport transport, byte[]... contents) throws Exception
+    {
+        start(transport, new HttpServlet()
```
If you store the reader thread like this:

```java
AtomicReference<Thread> readerThreadRef = new AtomicReference<>();
start(transport, new HttpServlet()
{
    @Override
    protected void service(HttpServletRequest request, HttpServletResponse response) throws IOException
    {
        readerThreadRef.set(Thread.currentThread());
        // Send 100-Continue and copy the content back
        IO.copy(request.getInputStream(), response.getOutputStream());
    }
});
```

then...

lorban commented on jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/HttpClientContinueTest.java:
```diff
@@ -111,6 +112,68 @@ protected void service(HttpServletRequest request, HttpServletResponse response)
+                    try
+                    {
+                        // Sleep here to ensure that sometimes the reader will block/pause even after sending 100.
+                        Thread.sleep(100);
```
...that `sleep` can be replaced with:

```java
await().atMost(5, TimeUnit.SECONDS).until(() ->
{
    Thread thread = readerThreadRef.get();
    if (thread == null)
        return false;
    return thread.getState() == Thread.State.WAITING;
});
```

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12/10956/Expect100Continue

gregw commented:
@sbordet @lorban can you just review the parts of this that are extra to #10957 

@sbordet I've made changes to the h2 server side to do the reset(NO_ERROR), but I've not yet changed the client side to handle it.  Not exactly sure what we should do there (if anything).  Need to chat.

@lorban can you look at duplicating the h2 changes in h3 once we agree they are correct.

gregw committed:
updates from review

gregw committed:
updates from review

Moved all 100 continues handling into HttpChannelState and protected it from concurrent writes

gregw committed:
Merge branch 'fix/jetty-12/10956/Expect100Continue' into fix/jetty-12/110956/Reset_NO_ERROR

# Conflicts:
#	jetty-core/jetty-http2/jetty-http2-server/src/main/java/org/eclipse/jetty/http2/server/internal/HttpStreamOverHTTP2.java

gregw commented:
there is a better way.
